### ISSUE TITLE
Session Option that is both Private/Public

### ISSUE THREAD

**Issue by TehTurk** [Neos-Metaverse/NeosPublic#2496]
## Is your feature request related to a problem? Please describe.
Following #2495 , I'm asking if there is a way to have a session be public, but only accessible by the link or private in some way. Allowing for things such as Metadata, or other world interactions for smoother transitions or events. 

## Relevant issues
#2495 #2494 #2493 

## Describe the solution you'd like
Currently Hidden removes all data, folks see this terminology as "Oh! It's not on the world browser" and treat it as a Public Session but *secret* Ideally a session setting or enhancing our current options to allow folks to join like a public session but only by links or otherwise. Or some form of toggle as you stated. 

## Describe alternatives you've considered
Setting stuff manually kinda works but is irksome and troublesome.

**Comment by Frooxius** [comment]
As I explained in the other issue, it can't really be both Private and Public, those are mutually exclusive.

From what I understand, you just want it to not be directly accessible through the world list, but rather just a hub?

I think it might be best to have this as headless-only option, since that's pretty much exclusively where it would be used when setting up hub worlds and similar.

The session would still be part of public metadata, it just wouldn't be shown in the browser list. However it's still technically public session and doesn't offer any actual privacy, so we need to avoid that terminology.

We could potentially call it a "Nested Session", since it's only going to be accessible from another one. We could even go a step further and specify where is the session supposed to be nested - e.g. world URL or session URL. This would mean that if you're currently in that world, those sessions will pop up in the world browser (as well as being available through worlds orbs), but not otherwise.

Would something like that work?

**Comment by Aerizeon** [comment]
That seems like a pretty good solution  - I definitely agree with making it a headless-only option.

**Comment by TehTurk** [comment]
Yeah I follow. Essentially kinda a hub yeah. 

Hmm I like the concept, but would it be also expandable for groups as well or is that too much? I'd figure if we are getting a direct implementation I'd also try to save you work where this overlaps as well.

Because lets say you are in that hub, or apart of that hubs group and want to see those sessions without connecting. (To do this you'd have to connect to check)

It'd allow folks to see sessions that are oriented with groups they are only part of as well. Allowing for groups to also have more group only sessions/communities. I know CloudVars would deal with this, but it also let folks see sessions related to a groups alot more easier too. (VBLFC)

If this is too much of an ask ignore my suggestion then just a thought I had when you brought up the idea and wanted to see if that'd be ideal  I'm fine with your idea 100% going forward.

**Comment by Frooxius** [comment]
What do you mean by groups in this context? As in sessions that are only accessible/visible to members of a particular group?

**Comment by TehTurk** [comment]
Yes, I figured if we got Nested sessions, letting group members see those nested sessions could be ideal as well. (If they were related, I don't know if the underlying URLs could accompany this)

**Comment by TehTurk** [comment]
As I said, if the idea is confusing or I'm not clear on how it could be an option feel free to ignore it as I don't want to detract from the main issue. It was something I wanted to bring up to save you time and resources as I know folks would want to join those public Sessions (but not join the hub to see it)

**Comment by Frooxius** [comment]
That'd be completely separate feature/filtering option. Nested sessions will still be public by themselves by default, you don't need to be member of any group to see them, just be in the right world.

Having sessions that are scoped to groups is also something we want to add at some point, but that'd be separate use-case.

**Comment by TehTurk** [comment]
Understood! I look forward to this :)

**Comment by TehTurk** [comment]
Ooh I didn't think of that. I do know there is a bit of issue of player/session split when doing some events so it does affect folks interest in it. This would be a neat addition but don't know if that'd be doable.

**Comment by shadowpanther** [comment]
> Will it be possible for the "hub" world to reflect the total user count/user list of all nested worlds as well?

This sounds like a by-room culling system, but instead having a separate session for each room to further lower the overhead. I could see it being a very useful system for large-scale events.

**Comment by Frooxius** [comment]
> 
> 
> Will it be possible for the "hub" world to reflect the total user count/user list of all nested worlds as well?

Potentially yes, at least with the way I'm considering to implement this, but question is, do we want it to be? It can be somewhat misleading.

Essentially what I'm thinking is that you'd add something like this:

```
parentSessions : 
[
"session A",
"session B"
]
```

This will make the session appear only when you're in either session A or session B. The data is always available though, so if the world orb or another portal or something is present elsewhere, it'll still show up.

If you don't want the session to show at all in the world listing and just use portals/world orbs that were manually placed, you'd just provide an empty list:

parentSessions: []

With this, we could simply just sum up all active users of each session that has "session A" when calculating its total nested number of users.

**Comment by Aerizeon** [comment]
@Frooxius I believe that in past experiments by medra and turk, a hub world showing no users tends to be get ignored, resulting in fairly low turnout. 

Allowing them to be combined (or at least reported on the hub world's listing in some manner) would allow for a more accurate representation of the number of users participating in an event, rather than depending on users to remain connected in the hub in the background.

**Comment by TehTurk** [comment]
If the goal for it is to be a hub world in some form I think that's fine then. As yeah Ep pointed out I and Medra run into Hub worlds be a bit of a mixed bag as this allows for folks to congregate alot easier but the numbers not being there cause issues. Big Number = More People for folks.

I really like how this is setup :o would those Sessions be preset ahead of time with Session IDs? Or can add or remove Nested Sessions to the main instance.

**Comment by TehTurk** [comment]
I just realized I misread this so let me rephrase. I really like this because it basically lets you have predefined lobby/session and then you can have any nested session basically *become* the child session of that lobby within the config. This males it really nice for meetups and I bet this will help bigger events as well, or folks wanting to do house parties too. 

If it's headless only it'd be neat to at least have a command to define a parent session as well too if that's the case.

**Comment by Frooxius** [comment]
> 
> 
> @Frooxius I believe that in past experiments by medra and turk, a hub world showing no users tends to be get ignored, resulting in fairly low turnout.
> 
> Allowing them to be combined (or at least reported on the hub world's listing in some manner) would allow for a more accurate representation of the number of users participating in an event, rather than depending on users to remain connected in the hub in the background.
> 
> I like the suggested implementation so far though :3

Ah I see, that makes sense! I'll probably split up the number of users internally, so we can tell how  many users are in the actual session and how many are overall.

@TehTurk Yeah. Technically you could use the temporary session ID's, but then you'd have to be updating your configs all the time, so the persistent ones will pretty much be the only practical choice. 

However there is one potential issue with this setup - technically anyone can mark their session as the child session of one of your sessions.

I'm considering if we should instead do the opposite - the parent session will specify all of its child sessions instead. That way you have control over all the ones that are included, but you'll have to update the config centrally if you want to add/remove more sessions, which makes it a bit harder.

It also makes it harder to make a nested session that technically doesn't have any parents where it should be displayed, but perhaps that could be separated out.

It would then become something like this:

```nestedSessions: 
[
"Session-A",
"Session-B",
"Session-C"
],
showNestedSessionsInBrowser: false```

But there are the two main issues with that:
- Adding/removing instances requires updating the hub session, so it doesn't scale as well
- When the hub session isn't running or visible, the nested sessions will all become root sessions instead. This might be even potentially desirable, but there's no way to control that with this approach.

**Comment by TehTurk** [comment]
Hmm both setups might be ideal depending. 
Because I run into situations where we need to spin up another instance on the fly and would be inflexible.

But for static worlds I can see the benefit of this approach. 

I had the idea Maybe just have a simple key for those children and if mismatch to parent then they become a root session? It adds some security but also authorization and keys could be remade on the fly config wise. But don't know if it's feasible in this way.

**Comment by Frooxius** [comment]
Hmm, I think having child sessions specify their parent sessions is probably the best solution for scalability and control, so I'll probably implement that one.

Then we could do an extra mechanism on top so parent sessions can control which child sessions will show up in the world browser - e.g. filtering it by the hosted worlds and/or the host account to restrict it - although there's also issue whether the filtering should be on by default or not. Meaning you need to setup filter in advance if you want any nested sessions to show.

**Comment by Frooxius** [comment]
> 
> 
> Hmm both setups might be ideal depending.
> Because I run into situations where we need to spin up another instance on the fly and would be inflexible.
> 
> But for static worlds I can see the benefit of this approach.
> 
> I had the idea Maybe just have a simple key for those children and if mismatch to parent then they become a root session? It adds some security but also authorization and keys could be remade on the fly config wise. But don't know if it's feasible in this way.

The  key would be quite more convoluted. It would need to be a digital signature, because it needs to be public. If it was just plain key, anyone would just be able to copy it. Instead the session would have to public key (e.g. RSA) and then each child session would sign itself (e.g. it's ID) with the private key. The digital signature would be included as part of the public session data, letting anyone to verify it with the parent session's public key.

It's possible, just a lot more complicated process.

**Comment by TehTurk** [comment]
Hmm I see, the key I was more of the mind a simple string for people to use and set ahead of time (and be only accessible within the config) and it be a tag check in a way. 

I'm not against the filter, but the problem rn with those two is most headless are typically best to run in 1 world per 1 account/setup as hosting more worlds doesn't scale network wise too cleanly. (Plus when Host isn't focused you run into issues too) So you run into alot of different accounts hosting, or alot of different hosts. 

I do like both filters and will work as options, it just doesn't answer the problems I run into usually I want to note. That and accounts can't be logged into on Multiple Accounts? So that's a thing that pops up for me along with this too if those were options and people wanted to actively use this.

**Comment by Frooxius** [comment]
@TehTurk Unfortunately the simple string won't work. If it's only in the config and not part of the session data, then others have no way to verify it.

I think if the filtering is based on the accounts and/or hosted world that can work well. For increased security we could do the RSA key one, but that will require more tooling and setup on user's end.

Being able to use same account across multiple instances of headless/Neos is planned, that's a separate issue: https://github.com/Neos-Metaverse/NeosPublic/issues/1427

**Comment by TehTurk** [comment]
I see. Thanks for at least explaining c:

No I agree, those are good options but it doesn't help with the problem I run into. I'd be fine running it in no filter mode but then I can get someone kinda just being a shit about it hosting sessions apart of an event I'm not too keen on.
Do you have any other ideas? Maybe add more filter types to allow more flexibility with it? Oh I know! Contacts Filtering? Might be be bit weird/awk to setup (but it's a preset setup situation) but if a bot has another bot friended or user friended, It would solve some of the concerns I think. Allowing some sort of stronger filter.

No I know! I'm bringing these points up at least so your aware.

**Comment by Frooxius** [comment]
Why exactly it doesn't help with the problem you run into? Can you elaborate on that?

What exactly do you mean by contacts filtering in this context? I'm not sure I understand how would you want that one to work.

**Comment by TehTurk** [comment]
Currently, for when we are hosting Meetups (where I can speak from) and for some of the other multi session events. If you want to host a headless session the ideal has always been 1 session, 1 host to reduce any sort of impact on the host, or how Neos handles the hosting. Because a headless host doesn't handle the scale of the data or most servers just don't handle the data in a way that doesn't overload a hosts connection. (Like network data out/in just gets to be too much or amount of bandwidth per user gets to be like 1.5 to 2.5 as more people are in the session or more stuff is going on) This also lets us do higher user numbers as 1 only need to worry about 1 connection point.

When others sessions are hosted on the same user, network events in other sessions hosted on the same user affect the other sessions and effect session health by proxy of having to deal with that amount of information.

So with these filters, it doesn't really let us do that approach and are locked to using the only 1 host approach or 1 account for Nested sessions/hubs. Meaning we can't have multiple users/accounts to spread the network load amongst the different hosts.

Ah! I mean where if the Bot/Headless Account is required to have that nested session added for it to be considered apart of the Hub . (As a Filter Type)

**Comment by Frooxius** [comment]
The solution I proposed doesn't have limitation to 1 host, you could specify as many as you want/need.

**Comment by shadowpanther** [comment]
But adding a completely new host into the event would require changing the config and restarting the "hub" server?
While filtering by the hub's contacts would allow to (cumbersomely) add a new contact and spin up a new nested session without restarting the hub.
And we already have contacts+cloudvars solution for controlling session access, maybe the same mechanism could be used.

**Comment by TehTurk** [comment]
Hmm I see. Yeah Shadow makes good points as well too for some other options.

**Comment by Frooxius** [comment]
You'd need to change the config yeah. Restart might not be necessary, we might potentially solve that by having a command to do it or having a way to reload those changes.

Using contacts like that isn't possible though. The mechanism for Contacts+ requires the user actually joining the session and the fetch is done by the host, rather than the joining user. This means it won't work here unfortunately. We'd have to introduce another mechanism and that's potentially a lot of complex engineering to do it securely.

However one alternative is for the host to actually manage the list of the sessions themselves. This would allow that to be used, although it would change the mechanism a bit.

This means that each session would "claim" to be a nested session of another one. The hub session would then see this and say "Yes, that session is mine" and put it into its metadata.

I think that can actually work and give a fair more flexibility!

**Comment by TehTurk** [comment]
Hmm I see. Maybe I'm not clear on how the sessions are interacting or like how the data is being shared so I thought this was possible in some form apologies.

Hmm my wonder is then then if the host is managing it does that mean we need to manually add each host to the config? I really liked the concept of the child defining the parent as it made it really easy to scale (1 ID, many sessions for other session configs) but with the list, each child needs to be specified.

**Comment by shadowpanther** [comment]
Yes, we either have
- to deal with impostors claiming they are nested children of our hub session
or
- to control which sessions _are_ our nested children by the host getting info for all claimants and then deciding which sessions to allow to be nested under

To do the second one, we need some form of a filter, either a predefined set of hosting users (in the config, requires restarting currently, might be reloaded when implemented) or by filtering by some definable property, like nested host being friends with the hub host (the one I suggested).

**Comment by TehTurk** [comment]
Well Froox brought up the concept where they "both" need to or can define each other. I think if the host had no filtering, this would be essentially the same where the Children can define the parent. So currently

- Host can Define (config setup)
- Children can Define (config setup, but easier to spin up on the fly)
- Both need to Define (pre set agreement but is a controlled situation and is expected)

My previous question was more to make the spinning up sessions easier with this in mind to work with the filter. 

Using an Account's friend list would seem more manageable from a user perspective and requires less overhead but that might require more work than currently needed at this time as pointed out by Froox.

**Comment by Frooxius** [comment]
Added nested sessions in 2021.6.18.4! I ended up doing the filtering on contacts, since it happens on each headless hosting the parent session now!

```
- Added support for nested sessions (based on request by @Turk | Mentor Manager, @Gawdl3y, @Epsilion and @Shadow Panther [RU/EN, UTC+3])
-- Each session can have one or more parent session IDs by specifying parentSessionIds in the headless config
-- Nested session won't show up in the world browser unless you are currently present in one of the parent sessions
--- IMPORTANT: This is not a security mechanism, but a cosmetic, the session and its metadata is still public even if you're not in the parent session
-- The host of the parent session must be a contact of the host of the nested session
-- Parent sessions now track total number of joined and active users in their child sessions and display this number in the world browser (this is useful for hub worlds)
-- SessionInfoSource now has TotalJoinedUsers and TotalActiveUsers
-- WorldListManager now has ParentSessionId property which allows showing sessions based on given parent session ID
--- The dash world browser has this driven to the SessionID of the currently focused world
-- This mechanism is mainly intended for public sessions. Using it with Contacts or Private sessions won't work properly
```
